### ISSUE TITLE
Add a programmatic way to get recipes

### ISSUE THREAD

**Issue by xsebek** [swarm-game/swarm#996]
It would be neat to be able to query the recipe requirements in Swarm language.

This would simplify creating an automated factory where robots are sent to fetch recipes dynamically.

Another use case is writing a helper function to list all requirements for a recipe:

```
>>> showMeWhatINeed "branch"
- 2 branch:
  - 1 tree
```

## Solution

I would like to have a command like this one:
```
recipe: entity -> cmd (entity * CmdFold (int * entity))

type CmdFold a = (a -> cmd unit) -> cmd unit
```

The `CmdFold` is basically a substitution for a list and can be inlined.

### Manual solution

It's certainly possible to write this function by hand, but it will get outdated.

### Scripted solution

A function generated by a script is easier to keep up to date but still would not work in challenges that add new recipes.

Also, it takes away from the fun of discovering recipes.

### Returning a list

This would be definitely more useful as the result could be used in a pure context. E.g. not take extra ticks in `atomic`.

**Comment by byorgey** [comment]
What should this command do if you run it on an entity you don't "know" yet?  Should it throw an exception?

**Comment by kostmo** [comment]
I'd like to understand the `CmdFold` idea more.  Does this require functionality that is currently missing?  Does it relate to the `cons`/`nil` technique demonstrated [here](https://github.com/swarm-game/swarm/pull/1214/files#r1174451625)?

Another approach for iterating list members, which seems to work well enough, is that used by the [`waypoint` and `structure` commands](https://github.com/swarm-game/swarm/blob/f9ef0094f0bfbe57d589e44a355446a54566d4bd/src/Swarm/Language/Typecheck.hs#L776-L777).  That is, take an index as input, and return a tuple of `(list length, item at index)`.

An API for recipe members might also be used for querying the materials required for structures (#1579).

**Comment by byorgey** [comment]
A `CmdFold a = (a -> cmd unit) -> cmd unit` represents a list of `a`s, in the sense that you can hand it any function of type `(a -> cmd unit)` you want, and the only thing it can do is call your function on some collection of `a` values and sequence the resulting commands.  So if you want to *e.g.* run some command for each ingredient of a recipe you would get the `CmdFold (int * entity)` and then pass it your `(int * entity -> cmd unit)` function, and your function would get run on each ingredient.

Actually, thinking about it now, I don't think `CmdFold` is general enough.  For example, what if you wanted to add up the total number of ingredients for a recipe?  If all you can do is have a `cmd unit` for each ingredient, adding up the counts would require some kind of mutable variables in the `cmd`  monad, which we don't have yet.

More general would just be a normal list fold.  So something like 

`recipe : entity -> cmd ((int * entity -> b -> b) -> b -> b)`

You (the caller of `recipe`) can choose the type `b` to be whatever you want.  It could be `cmd unit` and the combining function could just do sequencing.  But you could also choose `b` to be `int` (or `cmd int`, or whatever).  For example to add up the total ingredient count needed you could write

`r <- recipe "foozle"; return (r (\ingr. \cnt. cnt + fst ingr) 0)`

**Comment by byorgey** [comment]
Note that `forall b. (a -> b -> b) -> b -> b` is isomorphic to `List a`.  Showing this is a nice exercise in Haskell.

**Comment by byorgey** [comment]
Of course, now that we have lists (#1894) a lot of the above discussion becomes moot.  We could now have something like
```
recipe : Entity -> List (Int * Entity)
```
where `tydef List a = rec l. Unit + a * l end; tydef Entity = Text end`.

However, this omits a lot of subtlety about recipes, and I am not sure how we want to balance ease of use vs. capturing absolutely everything.  For example, some recipes require devices as a catalyst; some recipes can produce multiple outputs; and so on.  Maybe we could provide several functions ranging from a simple one like the above to more complex ones that capture more detail.  The `Recipe` data type currently looks like this:
```
data Recipe e = Recipe
  { _recipeInputs :: IngredientList e
  , _recipeOutputs :: IngredientList e
  , _recipeCatalysts :: IngredientList e
  , _recipeTime :: Integer
  , _recipeWeight :: Integer
  }
```
so capturing that fully in swarm-lang might look like
```
tydef Entity = Text end
tydef Ingredient = Int * Entity end
tydef Recipe = List Ingredient * List Ingredient * List Ingredient * Int * Int end

recipesProducing : Entity -> List Recipe
```

**Comment by byorgey** [comment]
Upgrading this to 'Moderate' since I think it's going to become indispensable as we continue to add more entities/devices, for the game to remain fun even as you get more advanced and have to deal with lots of different devices.

Another question is, what device should provide the ability to get recipes?  Some kind of `recipe book`?  Made of `paper` and... some other stuff?
